src/lib/operators/set_union.cpp

```diff
@@ -0,0 +1,371 @@
+      const auto segment_column_id_end = pos_lists_idx >= _column_segment_begins.size() - 1
+                                             ? _input_table_left()->column_count()
+                                             : _column_segment_begins[pos_lists_idx + 1];
+      for (auto column_id = segment_column_id_begin; column_id < segment_column_id_end; ++column_id) {
+        auto ref_column = std::make_shared<ReferenceColumn>(
+            _referenced_tables[pos_lists_idx], _referenced_column_ids[column_id], pos_lists[pos_lists_idx]);
+        chunk.add_column(ref_column);
+      }
+    }
+
+    out_table->emplace_chunk(std::move(chunk));
+  };
+
+  // Comparator of two ReferenceMatrix rows
+  const auto cmp = [](const auto& matrix_a, const auto idx_a, const auto& matrix_b, const auto idx_b) {
+    for (size_t column_idx = 0; column_idx < matrix_a.size(); ++column_idx) {
+      if (matrix_a[column_idx][idx_a] < matrix_b[column_idx][idx_b]) return true;
+      if (matrix_b[column_idx][idx_b] < matrix_a[column_idx][idx_a]) return false;
+    }
+    return false;
+  };
+
+  size_t chunk_row_idx = 0;
+  for (; left_idx < num_left || right_idx < num_right;) {
+    /**
+     * Begin derived from std::union()
+     */
+    if (left_idx == num_left) {
+      emit_row(reference_matrix_right, virtual_pos_list_right[right_idx]);
+      ++right_idx;
+    } else if (right_idx == num_right) {
+      emit_row(reference_matrix_left, virtual_pos_list_left[left_idx]);
+      ++left_idx;
+    } else if (cmp(reference_matrix_right, virtual_pos_list_right[right_idx], reference_matrix_left,
+                   virtual_pos_list_left[left_idx])) {
+      emit_row(reference_matrix_right, virtual_pos_list_right[right_idx]);
+      ++right_idx;
+    } else {
+      emit_row(reference_matrix_left, virtual_pos_list_left[left_idx]);
+
+      if (!cmp(reference_matrix_left, virtual_pos_list_left[left_idx], reference_matrix_right,
+               virtual_pos_list_right[right_idx])) {
+        ++right_idx;
+      }
+      ++left_idx;
+    }
+    ++chunk_row_idx;
+    /**
+     * End derived from std::union()
+     */
+
+    /**
+     * Emit a completed chunk
+     */
+    if (chunk_row_idx == out_chunk_size && out_chunk_size != 0) {
+      emit_chunk();
+
+      chunk_row_idx = 0;
+      std::generate(pos_lists.begin(), pos_lists.end(), [&] { return std::make_shared<PosList>(); });
+    }
+  }
+
+  if (chunk_row_idx != 0) {
+    emit_chunk();
+  }
+
+  return out_table;
+}
+
+std::shared_ptr<const Table> SetUnion::_analyze_input() {
+  Assert(_input_table_left()->column_count() == _input_table_right()->column_count(),
+         "Input tables must have the same layout. Column count mismatch.");
+
+  // Later code relies on input tables containing columns
+  if (_input_table_left()->column_count() == 0) {
+    return _input_table_left();
+  }
+
+  /**
+   * Check the column layout (column names and column types)
```

Comment:
This looks like something that could be moved into the table and be reused by Union*, Difference, ...